src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+        {
+            _workspace = workspace;
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+
+            Id = id;
+            Language = language;
+            _displayName = projectUniqueName;
+            _projectUniqueName = projectUniqueName;
+
+            if (directoryNameOpt != null)
+            {
+                _documentFileChangeContext = _workspace.FileChangeWatcher.CreateContextForDirectory(directoryNameOpt);
+            }
+            else
+            {
+                _documentFileChangeContext = workspace.FileChangeWatcher.CreateContext();
+            }
+
+            _documentFileChangeContext.FileChanged += FileChangeContext_FileChanged;
+
+            // TODO: set this to watch the NuGet directory or the reference assemblies directory; since those change rarely and most references
+            // will come from them, we can avoid creating a bunch of explicit file watchers.
+            _fileReferenceChangeContext = workspace.FileChangeWatcher.CreateContext();
+
+            _sourceFiles = new DocumentCollection(this, (s, d) => s.ContainsDocument(d), (w, d) => w.OnDocumentAdded(d), (w, documentId) => w.OnDocumentRemoved(documentId));
+            _additionalFiles = new DocumentCollection(this, (s, d) => s.ContainsAdditionalDocument(d), (w, d) => w.OnAdditionalDocumentAdded(d), (w, documentId) => w.OnAdditionalDocumentRemoved(documentId));
+        }
+
+        public ProjectId Id { get; }
+
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, Action<Workspace> changeValue)
+        {
+            lock (_gate)
+            {
+                // If nothing is changing, we can skip entirely
+                if (object.Equals(field, newValue))
+                {
+                    return;
+                }
+
+                field = newValue;
+
+                if (_activeBatchScopes > 0)
+                {
+                    _projectPropertyModificationsInBatch.Add(withNewValue);
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(changeValue);
+                }
+            }
+        }
+
+        private void ChangeProjectOutputPath(ref string field, string newValue, Func<Solution, Solution> withNewValue, Action<Workspace> changeValue)
+        {
+            lock (_gate)
+            {
+                // Skip if nothing changing
+                if (field == newValue)
+                {
+                    return;
+                }
+
+                if (field != null)
+                {
+                    _workspace.RemoveProjectOutputPath(Id, field);
+                }
+
+                if (newValue != null)
+                {
+                    _workspace.AddProjectOutputPath(Id, newValue);
+                }
```

Comment:
Isn't this a bit inconsistent? When batching ApplyChangeToWorkspace won't be called, but these will. Perhaps it doesn't matter here but perhaps it might be nice to have an invariant that workspace doesn't change until all batch scopes are closed.

Reply:
Yeah, you're right. I thought I had a reason for this, but I don't recall now. Moving that might help some lock contention.